pysmt/optimization/optimizer.py

```diff
@@ -0,0 +1,703 @@
+                    True: (mgr.Real, mgr.GT, mgr.GE),
+                    False: (mgr.Real, mgr.GT, mgr.GE),
+                },
+            },
+            BV: {
+                MinimizationGoal: {
+                    False: (cast_bv, mgr.BVULT, mgr.BVULE),
+                    True: (cast_bv, mgr.BVSLT, mgr.BVSLE),
+                },
+                MaximizationGoal: {
+                    False: (cast_bv, mgr.BVUGT, mgr.BVUGE),
+                    True: (cast_bv, mgr.BVSGT, mgr.BVSGE),
+                },
+            },
+        }
+        options[QF_LIRA] = options[LRA]
+        return options[goal.get_logic()][goal.opt()][goal.signed]
+
+
+class OptSearchInterval(OptComparationFunctions):
+    """
+    Represents a search interval for optimization problems.
+
+    This class is used to manage the bounds and constraints during the optimization
+    process. It supports both linear and binary search strategies to refine the
+    search space and determine the optimal value for a given goal.
+
+    Attributes:
+        _obj: The optimization goal (e.g., minimization or maximization).
+        _lower: The lower bound of the search interval.
+        _upper: The upper bound of the search interval.
+        _pivot: The current pivot value used in binary search.
+        environment: The solver environment.
+        client_data: Additional data used during the optimization process.
+    """
+
+    def __init__(self, goal, environment, client_data, lower, upper):
+        self._obj = goal
+        self._lower = lower
+        self._upper = upper
+        self._pivot = None
+        self.environment = environment
+        self._cast, self.op_strict, self.op_ns = self._comparation_functions(goal)
+        self.client_data = client_data
+
+    def linear_search_cut(self):
+        """
+        Generates a constraint for linear search.
+
+        This method creates a constraint to refine the search space by
+        cutting off part of the interval based on the current bounds.
+
+        Returns:
+            A constraint to refine the search space.
+        """
+        if self._obj.is_minimization_goal():
+            bound = self._upper
+        else:
+            bound = self._lower
+
+        return self.op_strict(self._obj.term(), self._cast(bound))
+
+    def _compute_pivot(self):
+        """
+        Computes the pivot value for binary search.
+
+        The pivot is calculated as the midpoint of the current bounds.
+
+        Returns:
+            The pivot value.
+        """
+        if self._lower is None and self._upper is None:
+            return 0
+        l, u = self._lower, self._upper
+        if self._lower is None and self._upper is not None:
+            l = self._upper - (abs(self._upper) + 1)
+        elif self._lower is not None and self._upper is None:
+            u = self._lower + abs(self._lower) + 1
+        add = 1 if self._obj.is_minimization_goal() else 0
+        return (l + u + add) // 2
```

Comment:
1. using a +1 is OK-ish for QF_LIA and QF_BV objectives, but not for QF_LRA (where we can simply use `/ 2` instead of '// 2')
2. I think the general rule is: if I am minimizing, then at each step an optimal solution `s` is s.t. `lower < s <= upper`. Instead, for maximization, `lower <= s < upper`. We have to document this. However, this means that if I have `lower=-10` and `upper=-9` in the case of maximization I would get a `pivot=-9` which would make the algorithm loop. I think a simpler approach would be to compute the pivot as `(l+u) // 2` and then adjust it with a +1 or -1 if it is outside of the ranges above.